### PR TITLE
Invoice update

### LINKED ISSUE #187: Order tasks alphabetically on invoice
There is a certain order in my project tasks, for instance:

01 Design
02 Build
03 Test
04 Deliver

I would like to have that exact order on my invoice (with sum hours), but at the moment the order is different. My request for change is to order the project tasks on the invoice alphabetically.
  
--
Hi vissersj,

I see that this can be of use. Please go to vissersj https://github.com/kimai/kimai/issues and file a feature request there.

Best Regards,
Severin

### LINKED ISSUE #468: Kimai invoice reverse task order
The standard task order for the kimai invoice extension starts with the most recent date and ends with the least recent date. 
My problem is that we need this the other way around, so start with the least recent task date and end with the most recent task date. So the task dates should be chronological. 

With older versions of the print.php i was able to reverse this, but so far i cant figure it out with the most recent version of kimai.

### PR DESCRIPTION
**Pull request by kevinpapst**
FIXES #468 
FIXES #187 

Changes proposed in this pull request:
- styling invoice extension screen
- removed rounding checkbox in favor of an empty option in rounding value select box (UI improvement)
- added failure handling for renderer
- setting customer contact data to empty string if not available
- added 3 select options for entry order: date-asc, date-desc, by name
- validating input parameter
- using own datepicker instead of relying on global timeframe
- new HTML invoice template: freelancer
- group invoice templates by type
- renamed existing invoice templates
- more fields available for rendering invoices
- rewritten and simplified code for future improvements

Changes within the core:
- reusing date_format_0 for datepicker (date_format_0 was not used anymore in Kimai)
- introducing a new date_format_3 setting to allow safely parsing datepicker values in PHP

This allows us to improve the datepicker usage in the future with different locales.

### CODE REVIEW

File: extensions/ki_invoice/init.php [hunk 1]
```diff
@@ -52,6 +52,7 @@
 
 // Select values for Round Time option
 $roundingOptions = array(
+    '0' => '',
     '1' => '0.1h',
```
Comment (simonschaufi) [inline extensions/ki_invoice/init.php:46]: I don't like the format of 0.1h. If you calculate 100 as a base (decimal format), the numbers have totally different meanings. This should be written as 1min, 30min..., then it is clear.

0.5h can also mean 1/2h = 30min but we mean 5min here or???. It even confuses me...

**Comment by simonschaufi** [comment]
nice layout! I :heart: it!

File: extensions/ki_invoice/templates/scripts/main.php [hunk 2]
```diff
@@ -6,45 +6,78 @@
...
+                        <?php echo $this->kga['lang']['ext_invoice']['defaultVat'] ?>
+                    </label>
+                    <span id="defaultVat"><?php echo $this->escape(str_replace('.', $this->kga['conf']['decimalSeparator'], $this->kga['conf']['defaultVat'])) ?></span>
+                    % (<a id="editVatLink" href="#"><?php echo $this->kga['lang']['change'] ?></a>)
+                </div>
+                <div>
+                    <label for="short">
+                        <?php echo $this->kga['lang']['ext_invoice']['invoiceOptionShort'] ?>
+                    </label>
+                    <input type="checkbox" name="short" checked="checked">
```
Comment (simonschaufi) [inline extensions/ki_invoice/templates/scripts/main.php]: id="short" missing to connect the label with the checkbox

File: extensions/ki_invoice/templates/scripts/main.php [hunk 3]
```diff
@@ -6,45 +6,78 @@
...
+                        <option value="-1" <?php if (!$this->kga['conf']['hideClearedEntries']): ?> selected="selected" <?php endif; ?>><?php echo $this->kga['lang']['export_extension']['cleared_all'] ?></option>
+                        <option value="1"><?php echo $this->kga['lang']['export_extension']['cleared_cleared'] ?></option>
+                        <option value="0" <?php if ($this->kga['conf']['hideClearedEntries']): ?> selected="selected" <?php endif; ?>><?php echo $this->kga['lang']['export_extension']['cleared_open'] ?></option>
+                    </select>
+                </div>
+                <div>
+                    <label for="sort_invoice">
+                        <?php echo $this->kga['lang']['ext_invoice']['sortOrder'] ?>
+                    </label>
+                    <select name="sort_invoice">
```
Comment (simonschaufi) [inline extensions/ki_invoice/templates/scripts/main.php]: id missing to connect the label with the select box

File: extensions/ki_invoice/templates/scripts/main.php [hunk 4]
```diff
@@ -6,45 +6,78 @@
...
+                    </label>
+                    <?php echo $this->formSelect('roundValue', null, array(
+                        'id' => 'invoice_round_ID',
+                        'class' => 'formfield'), $this->roundingOptions); ?>
+                </div>
+                <div>
+                    <label for="filter_cleared">
+                        <?php echo $this->kga['lang']['ext_invoice']['includedEntries'] ?>
+                    </label>
+                    <select name="filter_cleared">
```
Comment (simonschaufi) [inline extensions/ki_invoice/templates/scripts/main.php]: id missing to connect the label with the select box

File: extensions/ki_invoice/private_func.php [hunk 5]
```diff
@@ -200,4 +200,46 @@ function invoice_get_data($start, $end, $projects, $filter_cleared, $short_form)
...
+ * @return float
+ */
+function ext_invoice_round_value($value, $precision)
+{
+    // suppress division by zero error
+    if ($precision == 0.0) {
+        $precision = 1.0;
+    }
+
+    return floor($value / $precision + 0.5) * $precision;
```
Comment (simonschaufi) [inline extensions/ki_invoice/private_func.php:237]: can i has unit tests for this please?

**Comment by kevinpapst** [comment]
THIS IS STILL WORK IN PROGRESS.
Please do not merge until I comment again!

**Comment by kevinpapst** [comment]
Ready for review :-)
The rounding options need to be addressed, but I'd like to do that in a later stage, as I already changed (too) much in this one PR.

And you have some unit tests as well now!

**Comment by simonschaufi** [comment]
unit tests still fail because of undefined constant. please check travis for more info

**Comment by kevinpapst** [comment]
Argh, forgot to push the changed bootstrap.php - cool that we have unit tests ;-) even though only the tests are broken.
I will add that file tonight and then create a new pull request with cleaned commit history.

**Comment by simonschaufi** [comment]
no, please no new pull request. just make another commit and don't think about the history...

**Comment by kevinpapst** [comment]
Okay, here you go. Finalized PR

**Comment by simonschaufi** [comment]
I did some heavy testing and found the following problems:
- when I render the 2016 template, i get an error: `
  Fehler beim Rendern der Rechnung: item '<text:span' is not an existing key in the array.`
- when I render my own 2016 template, i get the following error: `Fehler beim Rendern der Rechnung: item 'fDuration' is not an existing key in the array.`

I'm not sure where the first error comes from but the second one has definitely something to do with this PR.
What exactly happened with 2016.odt? What did you change? I see the file size has changed but I don't see any visible changes. The "old" version still works fine.
I extracted the file to see the xml inside and found out you use LibreOffice on Mac. I use OpenOffice on Windows. Does this file work for you without problems?

**Comment by kevinpapst** [comment]
Thanks for your detailed testing and feedback! I love the new workflow and team work!!!!!
I will check everything and come back tomorrow.

**Comment by kevinpapst** [comment]
Okay, rechecked everything...
I removed the fDuration originally and renamed it to "duration", which would break the currently used and customized user-templates out there, so I re-added the field and added "duration" field with the duration in seconds.
The broken 2016 seems to come from the change of the fDuration field. I reverted it to your version, as my OO might be the reason for the break. 
Also added "start" and "end" fields as timestamps to fix #420.

You can re-test now.

**Comment by simonschaufi** [comment]
template variables should definitely also be documented. Please add this to the documentation as well

File: language/en.php [hunk 6]
```diff
@@ -453,12 +454,18 @@
...
         'invoiceOptionRound' => 'Round time',
         'invoiceButton' => 'Create',
-        'noData' => 'There are no time entries for the selected project & time period.',
+        'noData' => 'There are no entries for the selected project & time period',
         'noProject' => 'No project was selected.',
+        'includedEntries' => 'Include entries',
+        'failure' => 'Failed rendering invoice: %s',
+        'sortOrder' => 'Order',
+        'orderDateAsc' => 'Date - oldest first',
+        'orderDateDesc' => 'Datum - newest first',
```
Comment (simonschaufi) [inline language/en.php]: this should be "Date"

Comment (kevinpapst) [inline language/en.php]: Eagle-eye :+1:

**Comment by kevinpapst** [comment]
I will add them to the documentation later this day and send in a PR for that. Please merge them together then.

**Comment by simonschaufi** [comment]
@kevinpapst documentation of new variables is still missing!

**Comment by kevinpapst** [comment]
Yes, I need to test and add them to an invoice template first, thanks for the reminder!

**Pull request merged**